nodejs/aws-infra/experimental/clusterService.ts

```diff
@@ -0,0 +1,337 @@
+     */
+    public readonly autoScalingGroup?: mod.ClusterAutoScalingGroup;
+
+    constructor(type: string, name: string,
+                args: ClusterServiceArgs, isFargate: boolean,
+                opts: pulumi.ResourceOptions = {}) {
+        super(type, name, args, opts);
+
+        const parentOpts = { parent: this };
+
+        const dependsOn = opts.dependsOn
+            ? Array.isArray(opts.dependsOn) ? opts.dependsOn : [opts.dependsOn]
+            : [];
+
+        opts.dependsOn = dependsOn;
+
+        // If the cluster has an autoscaling group, ensure the service depends on it being created.
+        // TODO(cyrusn): this isn't necessary if resource creation automatically makes 'deps' for
+        // the opts passed in. Investigate.
+        if (args.autoScalingGroup) {
+            dependsOn.push(args.autoScalingGroup);
+        }
+
+        const loadBalancers = getLoadBalancers(this, name, args);
+
+        const clusterInstance = args.cluster;
+        const instance = new aws.ecs.Service(name, {
+            ...args,
+            loadBalancers,
+            cluster: clusterInstance.instance.arn,
+            taskDefinition: args.taskDefinition.instance.arn,
+            desiredCount: pulumi.output(args.desiredCount).apply(c => c === undefined ? 1 : c),
+            launchType: pulumi.output(args.launchType).apply(t => t || "EC2"),
+            waitForSteadyState: pulumi.output(args.waitForSteadyState).apply(w => w !== undefined ? w : true),
+            placementConstraints: pulumi.output(args.os).apply(os => placementConstraints(isFargate, os)),
+        }, parentOpts);
+
+        const taskDefinitionInstance = args.taskDefinition;
+        const autoScalingGroup = args.autoScalingGroup;
+
+        this.instance = instance;
+        this.clusterInstance = clusterInstance;
+        this.taskDefinitionInstance = args.taskDefinition;
+        this.autoScalingGroup = args.autoScalingGroup;
+
+        this.registerOutputs({
+            instance,
+            clusterInstance,
+            taskDefinitionInstance,
+            autoScalingGroup,
+        });
+    }
+}
+
+(<any>ClusterService).doNotCapture = true;
+
+function placementConstraints(isFargate: boolean, os: mod.HostOperatingSystem | undefined) {
+    if (isFargate) {
+        return [];
+    }
+
+    os = os || "linux";
+
+    return [{
+        type: "memberOf",
+        expression: `attribute:ecs.os-type == ${os}`,
+    }];
+}
+
+function getLoadBalancers(service: mod.ClusterService, name: string, args: ClusterServiceArgs) {
+    let loadBalancers: mod.LoadBalancers;
+    for (const containerName of Object.keys(args.taskDefinition.containers)) {
+        const container = args.taskDefinition.containers[containerName];
+        if (container.loadBalancerProvider) {
+            loadBalancers = combineLoadBalancers(
+                loadBalancers, container.loadBalancerProvider.loadBalancers(containerName, name, service));
+        }
+    }
+
+    return loadBalancers;
```

Comment:
another pattern.  Instead of populating the loadBalancers value itself.  The Service calls out to the .loadBalancerProvider to compute this based on the internal data it has.  This way the loadBalancerProvider can be implemented any way it wants to be.  It just needs to know how to return the 'loadBalancer' info that the Service needs.